lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -173,6 +173,73 @@ struct StorageGetOpLowering : public OpConversionPattern<arc::StorageGetOp> {
   }
 };
 
+struct AssertOpLowering : public OpConversionPattern<arc::AssertOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AssertOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+
+    rewriter.replaceOpWithNewOp<scf::IfOp>(
+        op, adaptor.getProperty(),
+        [&](auto &builder, auto loc) {
+          // if property is true then assert will not execute
+          builder.template create<scf::YieldOp>(loc);
+        },
+        [&](auto &builder, auto loc) {
+          // If the condition doesn't hold then print the message and abort
+          auto message = op.getMsg().has_value() ? op.getMsg().value() : "";
+          auto module = op->getParentOfType<ModuleOp>();
+
+          LLVM::LLVMFuncOp putsFunc;
+          LLVM::GlobalOp globalOp;
+
+          if (!message.empty()) {
```

Comment:
I think the `AssertOp` should have a verifier checking that if a message is given, it is not the empty string, otherwise we could just make it a default-valued attribute. This could then slightly simplify this lowering.